### ISSUE TITLE
CifCheck does not recognise "stop_"

### ISSUE THREAD

**Issue by bienchen** [rcsb/cpp-dict-pack#1]
According to the [Star format definition](https://www.iucr.org/resources/cif/spec/version1.1/cifsyntax), "stop_" is a reserved word and can not be used as unquoted data value. If I run CifCheck on the following snippet of mmCIF:

```data_C103531_G3_I1
_entry.id C103531_G3_I1
_struct.entry_id C103531_G3_I1
_struct.pdbx_model_details stop_in_the_name_of_love
_struct.pdbx_structure_determination_methodology computational
_struct.title 'Model'
```

CifCheck does not complain about `_struct.pdbx_model_details stop_in_the_name_of_love` but it should. When reading the above mmCIF input using the [RCSB Python mmcif package](https://pypi.org/project/mmcif/), I do get

```mmcif.io.PdbxExceptions.PdbxSyntaxError:  [Line: 4] Unexpected reserved word: stop```

So I guess CifCheck should also detect the reserved words in Star format.

Thanks,

Bienchen

**Comment by epeisach** [comment]
You are correct. The CIF specification indicates that "stop_" is a reserved keyword and should be quoted.
https://www.iucr.org/resources/cif/spec/version1.1/cifsyntax

What is unclear is if parsing should resume immediately after stop_ - without whitespace or not.

Which parser in mmcif.io are you using?  The core or pythonic adapter.  The mmcif package for C++ parser shares the same code as the dictionary. 

I will need to look into this.

**Comment by bienchen** [comment]
I'm using the `mmcif` Python package version 0.76, installed via pip, and then `PdbxReader` from `mmcif.io.PdbxReader`.

**Comment by epeisach** [comment]
Ok - PdbxReader use the purely python based parser.  So the deficiency is in the C++ used parser - which is shared by the mmcif and dictionary packages.

**Comment by epeisach** [comment]
Looking at the code - I also suspect PdbxWriter will not quote such a string - which violates spec as well.

**Comment by bienchen** [comment]
That is a good point. Haven't checked this since I can't read the particular file with PdbxReader to load something to write.

**Comment by josemduarte** [comment]
Interestingly, @piehld has recently encountered a related issue reading a CIF file from AlphaFoldDB. [This AlphaFoldDB entry](https://alphafold.ebi.ac.uk/entry/A4I106) has this:
```
#
_entity.details                  ?
_entity.formula_weight           ?
_entity.id                       1
_entity.pdbx_description         "STOP_axonemal_protein_-_putative"
_entity.pdbx_ec                  ?
_entity.pdbx_fragment            ?
_entity.pdbx_mutation            ?
_entity.pdbx_number_of_molecules 1
_entity.src_method               man
_entity.type                     polymer

```
And the parser apparently didn't like the `STOP_` string, even if quoted. @piehld would you have more details of parser and version?

**Comment by piehld** [comment]
I don't have details on the parser (the error is arising at [rcsb.utils.io.IoUtils](https://github.com/rcsb/py-rcsb_utils_io/blob/master/rcsb/utils/io/IoUtil.py), but I haven't dug further down yet). But reading all of your comments, I think what's happening is our internal parser is able to read that snippet fine (since it's quoted), but then when we are re-writing the data file back out, it becomes unquoted and thus leads to a parsing error.

**Comment by epeisach** [comment]
I have a fix for the python based writer to properly quote.  I need to track down the C++ based one.

**Comment by epeisach** [comment]
I have the code changes.  I am running through full test suite.
Would people like to commit a change that will fix the writer?

This is separate from the dictionary tools not picking up on invalid syntax when using the C++ reader.

**Comment by piehld** [comment]
Thanks @epeisach, are you able to share the link to your code changes?

**Comment by epeisach** [comment]
Not yet....  Still running tests. (quoting not working for me in comment)

Also have test suite changes as well.
I am pulling in latest GitHub modules at same time.
String::IsCiEqual is a case insensitive comparison

`
diff --git a/mmcif/api/DataCategoryFormatted.py b/mmcif/api/DataCategoryFormatted.py
index 9e4846d..8bf3123 100644
--- a/mmcif/api/DataCategoryFormatted.py
+++ b/mmcif/api/DataCategoryFormatted.py
@@ -133,7 +133,7 @@ class DataCategoryFormatted(DataCategory):
                     return (self.__doubleQuotedList(inp), "DT_ITEM_NAME")
                 elif inp[0] in ["[", "]", "$", "#", ";"]:
                     return (self.__doubleQuotedList(inp), "DT_DOUBLE_QUOTED_STRING")
-                elif inp[:5].lower() in ["data_", "loop_", "save_"]:
+                elif inp[:5].lower() in ["data_", "loop_", "save_", "stop_"]:
                     return (self.__doubleQuotedList(inp), "DT_DOUBLE_QUOTED_STRING")
                 else:
                     return ([str(inp)], "DT_UNQUOTED_STRING")
`
For C++ writer - under modules/cpp_cif_file/src/CifFile.C
diff --git a/src/CifFile.C b/src/CifFile.C
index 4cf1379..1305607 100644
--- a/src/CifFile.C
+++ b/src/CifFile.C
@@ -441,6 +441,8 @@ int CifFile::_IsQuotableText(const string& itemValue)
       return(1);
   if (String::IsCiEqual(itemValue.substr(0, 5), "save_"))
       return(1);
+  if (String::IsCiEqual(itemValue.substr(0, 5), "stop_"))
+      return(1);
 
   return(0);
 
@@ -647,6 +649,8 @@ int CifFile::_PrintItemValue(ostream& cifo, const string& itemValue,
         multipleWord = true;
     if (String::IsCiEqual(itemValue.substr(0, 5), "save_"))
         multipleWord = true;
+    if (String::IsCiEqual(itemValue.substr(0, 5), "stop_"))
+        multipleWord = true;
 
     if (embeddedQuotes && multipleWord)
         multipleLine = true;
@@ -884,6 +888,8 @@ int CifFile::_PrintItemNameInHeader(ostream& cifo, const string& itemValue,
         multipleWord = true;
     if (String::IsCiEqual(itemValue.substr(0, 5), "save_"))
         multipleWord = true;
+    if (String::IsCiEqual(itemValue.substr(0, 5), "stop_"))
+        multipleWord = true;
 
     if (embeddedQuotes && multipleWord)
         multipleLine = true;

**Comment by epeisach** [comment]
@piehld I put in a pull request

**Comment by piehld** [comment]
Thanks @epeisach! That seems like it should resolve all the writing behavior at least. Should we also allow for the reading of a non-quoted "stop_" string, or leave that functionality as is to maintain CIF specifications.
